### PR TITLE
RavenDB-14342

### PR DESCRIPTION
**Pull request by ml054**

### CODE REVIEW

**Comment by ml054** [comment]
Please review server side before I continue working on telegraf part. I've like to make sure endpoint outputs looks good etc.

File: src/Raven.Client/Exceptions/Commercial/LimitType.cs [hunk 1]
```diff
@@ -62,6 +62,9 @@ public enum LimitType
         EncryptedBackup,
 
         [Description("Additional Assemblies from NuGet")]
-        AdditionalAssembliesFromNuGet
+        AdditionalAssembliesFromNuGet,
+        
+        [Description("Endpoints Monitoring")]
```
Comment (ppekrol) [inline src/Raven.Client/Exceptions/Commercial/LimitType.cs]: Shouldn't this be the other way around? Monitoring Endpoints?

File: src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+                [nameof(Statistics)] = Statistics.ToJson(),
+                [nameof(Indexes)] = Indexes.ToJson(),
+                [nameof(Storage)] = Storage.ToJson()
+            };
+        }
+    }
+
+    public class DatabaseStatistics
+    {
+        public int DocPutsPerSecond { get; set; }
```
Comment (ppekrol) [inline src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs]: Maybe `PerSec` ?

File: src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    public class DatabaseCounts
+    {
+        public long Documents { get; set; }
+        public long Revisions { get; set; }
+        public long Attachments { get; set; }
+        public long UniqueAttachments { get; set; }
+        public long Alerts { get; set; }
+        public int Rehabs { get; set; }
```
Comment (ppekrol) [inline src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs:74]: What is this?

File: src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    public class DatabaseIndexesMetrics
+    {
+        public long Count { get; set; }
+        public int StaleCount { get; set; }
+        public long ErrorsCount { get; set; }
+        public int StaticCount { get; set; }
+        public int AutoCount { get; set; }
+        public int IdleCount { get; set; }
+        public int DisabledCount { get; set; }
+        public int ErrorCount { get; set; }
```
Comment (ppekrol) [inline src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs]: Errored maybe?

File: src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        }
+    }
+
+    public class DatabaseStorageMetrics
+    {
+        public long DocumentsAllocatedDataFileInMb { get; set; }
+        public long DocumentsUsedDataFileInMb { get; set; }
+        public long IndexesAllocatedDataFileInMb { get; set; }
+        public long IndexesUsedDataFileInMb { get; set; }
+        public long TotalAllocatedStorageFileInMb { get; set; }
```
Comment (ppekrol) [inline src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs:127]: In previous you have Data here you have Storage? What is the difference?

File: src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.Utils.Monitoring
+{
+    public class DatabaseMetrics
+    {
+        public string Name { get; set; }
```
Comment (ppekrol) [inline src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs]: DatabaseName?

File: src/Raven.Server/Utils/Monitoring/ServerMetrics.cs [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+                [nameof(ThreadPoolAvailableWorkerThreads)] = ThreadPoolAvailableWorkerThreads,
+                [nameof(ThreadPoolAvailableCompletionPortThreads)] = ThreadPoolAvailableCompletionPortThreads
+            };
+        }
+    }
+
+    public class MemoryMetrics
+    {
+        public long TotalMemoryInMb { get; set; }
+        public bool LowState { get; set; }
```
Comment (ppekrol) [inline src/Raven.Server/Utils/Monitoring/ServerMetrics.cs]: LowState?

I think we also have 3 states now, Low, ExtremlyLow, Normal?

File: src/Raven.Server/Utils/Monitoring/ServerMetrics.cs [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+        
+        public DynamicJsonValue ToJson()
+        {
+            return new DynamicJsonValue
+            {
+                [nameof(ServerVersion)] = ServerVersion,
+                [nameof(ServerFullVersion)] = ServerFullVersion,
+                [nameof(UpTimeInSec)] = UpTimeInSec,
+                [nameof(ServerProcessId)] = ServerProcessId,
+                [nameof(CurrentNumberOfRunningBackups)] = CurrentNumberOfRunningBackups,
```
Comment (ppekrol) [inline src/Raven.Server/Utils/Monitoring/ServerMetrics.cs]: I'm wondering if it is not better to wrap this in Backups and put there not only this but also max number of concurrent backups?

File: src/Raven.Server/Utils/Monitoring/ServerMetrics.cs [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+                [nameof(Certificate)] = Certificate.ToJson(),
+                [nameof(Cluster)] = Cluster.ToJson(),
+                [nameof(Databases)] = Databases.ToJson()
+            };
+        }
+    }
+
+    public class ConfigurationMetrics
+    {
+        public string[] Urls { get; set; }
```
Comment (ppekrol) [inline src/Raven.Server/Utils/Monitoring/ServerMetrics.cs]: maybe this should be moved to NetworkMetrics?

File: src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs (hunk 3 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs:74]: it matches this:

Database Stats | Number of rehabs | 1.3.6.1.4.1.45751.1.1.5.2.X.1.14

File: src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs (hunk 4 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs]: yes - this is better name - so errored mean amout of indexes with error state, and errors count is sum of indexing errors.

Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs]: I will describe each prop in telegaf input plugin readme file along with metric name.

File: src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs (hunk 5 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs:127]: total = documents storage + indexes storage + confguration storage (all storage envs)

File: src/Raven.Server/Utils/Monitoring/ServerMetrics.cs (hunk 8 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/ServerMetrics.cs]: ok

File: src/Raven.Server/Utils/Monitoring/ServerMetrics.cs (hunk 9 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/ServerMetrics.cs]: my idea was to put here items are not changing when server is running - (configuration)

in network we have items that change over time. So up to you.

File: src/Raven.Server/Utils/Monitoring/ServerMetrics.cs (hunk 7 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/ServerMetrics.cs]: ok will check that and replace with enum.

File: src/Raven.Client/Exceptions/Commercial/LimitType.cs (hunk 1 above)
Comment (ml054) [reply src/Raven.Client/Exceptions/Commercial/LimitType.cs]: changed

File: src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs (hunk 2 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs]: replacing all PerSecond -> PerSec

It will match InSec

Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs]: done

File: src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs (hunk 4 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs]: done

File: src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs (hunk 6 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs]: done

File: src/Raven.Server/Utils/Monitoring/ServerMetrics.cs (hunk 7 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/ServerMetrics.cs]: changed to: LowMemorySeverity

File: src/Raven.Server/Utils/Monitoring/ServerMetrics.cs (hunk 8 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/ServerMetrics.cs]: done

File: src/Raven.Server/Commercial/LicenseManager.cs [hunk 10]
```diff
@@ -1310,6 +1310,26 @@ public void AssertCanAddSqlEtl()
             const string message = "Your current license doesn't include the SQL ETL feature";
             throw GenerateLicenseLimit(LimitType.SqlEtl, message);
         }
+        
+        public bool CanUseMonitoringEndpoints(bool withNotification)
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/LicenseManager.cs:1314]: We also need to handle `ThrowIfCannotActivateLicense` method?

File: src/Raven.Server/Utils/Monitoring/CollectionMetrics.cs [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+                [nameof(DocumentsCount)] = DocumentsCount,
+                [nameof(TotalSizeInBytes)] = TotalSizeInBytes,
+                [nameof(DocumentsSizeInBytes)] = DocumentsSizeInBytes,
+                [nameof(TombstonesSizeInBytes)] = TombstonesSizeInBytes,
+                [nameof(RevisionsSizeInBytes)] = RevisionsSizeInBytes
+            };
+        }
+    }
+    
+    public class CollectionsMetrics
```
Comment (ppekrol) [inline src/Raven.Server/Utils/Monitoring/CollectionMetrics.cs:55]: Where is this used? How it is used? From design looks like we will have a list of all collections from all of the databases in one list?

File: src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+    public class DatabaseStatistics
+    {
+        public double DocPutsPerSec { get; set; }
+        public double MapIndexIndexesPerSec { get; set; }
+        public double MapReduceIndexMappedPerSec { get; set; }
+        public double MapReduceIndexReducedPerSec { get; set; }
+        public double RequestsPerSec { get; set; }
+        public int RequestsCount { get; set; }
+        public double RequestAverageDuration { get; set; }
```
Comment (ppekrol) [inline src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs]: InMs?

File: src/Raven.Server/Utils/Monitoring/ServerMetrics.cs [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+        
+        public DynamicJsonValue ToJson()
+        {
+            return new DynamicJsonValue
+            {
+                [nameof(ServerVersion)] = ServerVersion,
+                [nameof(ServerFullVersion)] = ServerFullVersion,
+                [nameof(UpTimeInSec)] = UpTimeInSec,
+                [nameof(ServerProcessId)] = ServerProcessId,
+                [nameof(Config)] = Config.ToJson(),
```
Comment (ppekrol) [inline src/Raven.Server/Utils/Monitoring/ServerMetrics.cs:40]: Can some of them be null?

File: src/Raven.Server/Web/System/AdminMonitoringHandler.cs [hunk 14]
```diff
@@ -0,0 +1,591 @@
...
+            result.Config = GetConfigMetrics();
+            result.Backup = GetBackupMetrics();
+            result.Cpu = GetCpuMetrics();
+            result.Memory = GetMemoryMetrics();
+            result.Network = GetNetworkMetrics();
+            result.License = GetLicenseMetrics();
+            result.Disk = GetDiskMetrics();
+            result.Certificate = GetCertificateMetrics();
+            result.Cluster = GetClusterMetrics();
+            result.Databases = GetAllDatabasesMetrics();
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/AdminMonitoringHandler.cs:60]: I'm a bit worried that there is no ability to say what should be returned. Can we add switches e.g. `?cluster=false` ?

File: src/Raven.Server/Web/System/AdminMonitoringHandler.cs [hunk 15]
```diff
@@ -0,0 +1,591 @@
...
+            result.UtilizedCpuCores = Server.ServerStore.LicenseManager.GetCoresLimitForNode(out _);
+            result.MaxCores = licenseStatus.MaxCores;
+            return result;
+        }
+
+        private DiskMetrics GetDiskMetrics()
+        {
+            var result = new DiskMetrics();
+            var environmentStats = Server.ServerStore._env.Stats();
+            result.SystemStoreUsedDataFileSizeInMb = environmentStats.UsedDataFileSizeInBytes / 1024L / 1024L;
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/AdminMonitoringHandler.cs]: use `Size` and `.GetValue`

File: src/Raven.Server/Utils/Monitoring/ServerMetrics.cs (hunk 13 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/ServerMetrics.cs:40]: no - we always collect that

File: src/Raven.Server/Web/System/AdminMonitoringHandler.cs (hunk 14 above)
Comment (ml054) [reply src/Raven.Server/Web/System/AdminMonitoringHandler.cs:60]: collecting all data is O(1) so I don't think we should worry about that. I just grouped them into categories.

File: src/Raven.Server/Commercial/LicenseManager.cs (hunk 10 above)
Comment (ml054) [reply src/Raven.Server/Commercial/LicenseManager.cs:1314]: ?

File: src/Raven.Server/Utils/Monitoring/CollectionMetrics.cs (hunk 11 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/CollectionMetrics.cs:55]: All db related endpoints follow the same structure: flat list with database / index or collection name:

image

I skipped grouping by database. Not sure this is needed.

File: src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs (hunk 12 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/DatabaseMetrics.cs]: ahhh I missed that - thanks!

File: src/Raven.Server/Web/System/AdminMonitoringHandler.cs (hunk 15 above)
Comment (ml054) [reply src/Raven.Server/Web/System/AdminMonitoringHandler.cs]: replaced all 7 usages.

File: src/Raven.Server/Utils/Monitoring/CollectionMetrics.cs (hunk 11 above)
Comment (ml054) [reply src/Raven.Server/Utils/Monitoring/CollectionMetrics.cs:55]: implemented grouping - just in case

**Review by ppekrol — APPROVED** [review]

**Pull request merged**